pkg/target/http_test.go

```diff
@@ -645,6 +645,156 @@ func TestHttpWrite_TLS(t *testing.T) {
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[1]}) //group 2
+}
+
+func TestHTTP_GroupByHeaders_Enabled_DifferentHeaderValue(t *testing.T) {
+	assert := assert.New(t)
+	target := HTTPTarget{dynamicHeaders: true}
+
+	inputMessages := []*models.Message{
+		{Data: []byte("value1"), HTTPHeaders: map[string]string{"h1": "v1"}}, //group 1
+		{Data: []byte("value2"), HTTPHeaders: map[string]string{"h1": "v2"}}, //group 2
+	}
+
+	groupedMessages := target.groupByDynamicHeaders(inputMessages)
+	assert.Len(groupedMessages, 2)
+
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[0]}) //group 1
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[1]}) //group 2
+}
+
+func TestHTTP_GroupByHeaders_Enabled_DifferentHeaderName(t *testing.T) {
+	assert := assert.New(t)
+	target := HTTPTarget{dynamicHeaders: true}
+
+	inputMessages := []*models.Message{
+		{Data: []byte("value1"), HTTPHeaders: map[string]string{"h1": "v1"}}, //group 1
+		{Data: []byte("value2"), HTTPHeaders: map[string]string{"h2": "v1"}}, //group 2
+	}
+
+	groupedMessages := target.groupByDynamicHeaders(inputMessages)
+	assert.Len(groupedMessages, 2)
+
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[0]}) //group 1
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[1]}) //group 2
+}
+
+func TestHTTP_GroupByHeaders_Enabled_AdditionalHeader(t *testing.T) {
+	assert := assert.New(t)
+	target := HTTPTarget{dynamicHeaders: true}
+
+	inputMessages := []*models.Message{
+		{Data: []byte("value1"), HTTPHeaders: map[string]string{"h1": "v1"}},                    //group 1
+		{Data: []byte("value2"), HTTPHeaders: map[string]string{"h1": "v1", "additonal": "v2"}}, //group 2
+	}
+
+	groupedMessages := target.groupByDynamicHeaders(inputMessages)
+	assert.Len(groupedMessages, 2)
+
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[0]}) //group 1
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[1]}) //group 2
+}
+
+func TestHTTP_GroupByHeaders_Enabled_MultipleGroups(t *testing.T) {
+	assert := assert.New(t)
+	target := HTTPTarget{dynamicHeaders: true}
+
+	inputMessages := []*models.Message{
+		{Data: []byte("value1")}, //group 1
+		{Data: []byte("value2")}, //group 1
+		{Data: []byte("value3"), HTTPHeaders: map[string]string{"h1": "v1"}}, //group 2
+		{Data: []byte("value4"), HTTPHeaders: map[string]string{"h1": "v1"}}, //group 2
+		{Data: []byte("value5"), HTTPHeaders: map[string]string{"h1": "v2"}}, //group 3
+	}
+
+	groupedMessages := target.groupByDynamicHeaders(inputMessages)
+	assert.Len(groupedMessages, 3)
+
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[0], inputMessages[1]}) //group 1
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[2], inputMessages[3]}) //group 2
+	assert.Contains(groupedMessages, []*models.Message{inputMessages[4]})                   //group 3
+}
+
+func TestHTTPWrite_GroupedRequests(t *testing.T) {
```

Comment:
Didn't think of it this ahead of time but this one lives somewhere between my PR for batch tests and this one.

You merge first, should be fairly easy for me to rebase on that.